src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -950,39 +1046,115 @@ bool CachedReadBufferFromRemoteFS::nextImplStep()
+        auto cache_file_path = getFileNameFromReadBuffer(*implementation_buffer);
 
         throw Exception(
             ErrorCodes::LOGICAL_ERROR,
-            "Having zero bytes, but range is not finished: file offset: {}, reading until: {}, read type: {}, cache file size: {}",
+            "Having zero bytes, but range is not finished: file offset: {}, starting offset: {}, "
+            "reading until: {}, read type: {}, cache file size: {}, cache file path: {}, "
+            "cache file offset: {}, current file segment: {}",
             file_offset_of_buffer_end,
+            first_offset,
             read_until_position,
             toString(read_type),
-            cache_file_size ? std::to_string(*cache_file_size) : "None");
+            cache_file_size ? std::to_string(cache_file_size) : "None",
+            cache_file_path,
+            implementation_buffer->getFileOffsetOfBufferEnd(),
+            file_segment->getInfoForLog());
     }
 
     return result;
 }
 
-off_t CachedReadBufferFromRemoteFS::seek(off_t offset, int whence)
+off_t CachedOnDiskReadBufferFromFile::seek(off_t offset, int whence)
 {
-    if (initialized)
-        throw Exception(ErrorCodes::CANNOT_SEEK_THROUGH_FILE, "Seek is allowed only before first read attempt from the buffer");
+    if (initialized && !allow_seeks)
+    {
+        throw Exception(
+            ErrorCodes::CANNOT_SEEK_THROUGH_FILE,
+            "Seek is allowed only before first read attempt from the buffer");
+    }
 
-    if (whence != SEEK_SET)
+    size_t new_pos = offset;
+
+    if (allow_seeks)
+    {
+        if (whence != SEEK_SET && whence != SEEK_CUR)
+        {
+            throw Exception("Expected SEEK_SET or SEEK_CUR as whence", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+        }
+
+        if (whence == SEEK_CUR)
+        {
+            new_pos = file_offset_of_buffer_end - (working_buffer.end() - pos) + offset;
+        }
+
+        if (new_pos + (working_buffer.end() - pos) == file_offset_of_buffer_end)
+            return new_pos;
+
+        if (file_offset_of_buffer_end - working_buffer.size() <= new_pos && new_pos <= file_offset_of_buffer_end)
+        {
+            pos = working_buffer.end() - file_offset_of_buffer_end + new_pos;
+            assert(pos >= working_buffer.begin());
+            assert(pos <= working_buffer.end());
+            return new_pos;
+        }
+    }
+    else if (whence != SEEK_SET)
+    {
         throw Exception(ErrorCodes::CANNOT_SEEK_THROUGH_FILE, "Only SEEK_SET allowed");
+    }
 
-    first_offset = offset;
-    file_offset_of_buffer_end = offset;
-    size_t size = getTotalSizeToRead();
-    initialize(offset, size);
+    first_offset = file_offset_of_buffer_end = new_pos;
+    resetWorkingBuffer();
 
-    return offset;
+    if (file_segments_holder && current_file_segment_it != file_segments_holder->file_segments.end())
+    {
+        // auto & file_segments = file_segments_holder->file_segments;
```

Comment:
I want to leave this commented code, because it should be a good improvement to reduce number of seeks, but it is very complex and need to be done very carefully and I will not do it right now, but having it there will remind that it needs to be done.
Or I can remove this commented code and add a TODO if it is better.